File: python/yacht/yacht.py
# Score categories.
# Change the values as you see fit.
YACHT = 'Yacht'
ONES = 1
TWOS = 2
THREES = 3
FOURS = 4
FIVES = 5
SIXES = 6
FULL_HOUSE = 'Full House'
FOUR_OF_A_KIND = 'Four of a Kind'
LITTLE_STRAIGHT = 'Little Straight'
BIG_STRAIGHT = 'Big Straight'
CHOICE = 'Choice'


def score(dice, category):
    dice.sort()
    if category == CHOICE:
        return sum(dice)
    elif category == YACHT:
        return (dice[0] == dice[4]) * 50
    elif category == FOUR_OF_A_KIND:
        return (dice[0] == dice[3] or dice[1] == dice[4]) * 4 * dice[3]
    elif category == FULL_HOUSE:
        return (3 * dice[0] + 2 * dice[4] if  dice[0] == dice[2] and dice[2] != dice[3] and dice[3] == dice[4]
        else 2 * dice[0] + 3 * dice[4] if dice[0] == dice[1] and dice[1] != dice[2] and dice[2] == dice[4]
        else 0)
    elif category == LITTLE_STRAIGHT:
        return (is_straight(dice) and dice[0] == 1) * 30
    elif category == BIG_STRAIGHT:
        return (is_straight(dice) and dice[0] == 2) * 30
    else:
        return len([die for die in dice if die == category]) * category


def is_straight(dice):
    return dice[0] + 1 == dice[1] and dice[1] + 1 == dice[2] and dice[2] + 1 == dice[3] and dice[3] + 1 == dice[4]

File: elixir/yacht/lib/yacht.ex
defmodule Yacht do
  @type category ::
          :ones
          | :twos
          | :threes
          | :fours
          | :fives
          | :sixes
          | :full_house
          | :four_of_a_kind
          | :little_straight
          | :big_straight
          | :choice
          | :yacht

  @doc """
  Calculate the score of 5 dice using the given category's scoring method.
  """
  @spec score(category :: category(), dice :: [integer]) :: integer
  def score(:ones, dice), do: Enum.count(dice, &(&1 == 1))
  def score(:twos, dice), do: 2 * Enum.count(dice, &(&1 == 2))
  def score(:threes, dice), do: 3 * Enum.count(dice, &(&1 == 3))
  def score(:fours, dice), do: 4 * Enum.count(dice, &(&1 == 4))
  def score(:fives, dice), do: 5 * Enum.count(dice, &(&1 == 5))
  def score(:sixes, dice), do: 6 * Enum.count(dice, &(&1 == 6))
  def score(:choice, dice), do: Enum.sum(dice)
  def score(:yacht, [a, a, a, a, a]), do: 50
  def score(:yacht, _), do: 0
  def score(category, dice) do
    case {category, Enum.sort(dice)} do
      {:little_straight, [1, 2, 3, 4, 5]} -> 30
      {:big_straight, [2, 3, 4, 5, 6]} -> 30
      {:full_house, [a, a, a, a, a]} -> 0
      {:full_house, [a, a, a, b, b]} -> 3 * a + 2 * b
      {:full_house, [a, a, b, b, b]} -> 2 * a + 3 * b
      {:four_of_a_kind, [a, a, a, a, _]} -> 4 * a
      {:four_of_a_kind, [_, a, a, a, a]} -> 4 * a
      _ -> 0
    end
  end
end
